crates/engine/tree/src/tree/payload_processor/prewarm.rs

```diff
@@ -71,41 +62,65 @@ where
         execution_cache: ExecutionCache,
         ctx: PrewarmContext<N, P, Evm>,
         to_multi_proof: Option<Sender<MultiProofMessage>>,
-        pending: VecDeque<Recovered<N::SignedTx>>,
-    ) -> Self {
+    ) -> (Self, Sender<PrewarmTaskEvent>) {
         let (actions_tx, actions_rx) = channel();
-        Self {
-            executor,
-            execution_cache,
-            pending,
-            ctx,
-            max_concurrency: 64,
-            to_multi_proof,
-            actions_rx,
+        (
+            Self {
+                executor,
+                execution_cache,
+                ctx,
+                max_concurrency: 64,
+                to_multi_proof,
+                actions_rx,
+            },
             actions_tx,
-            prewarm_outcomes_left: 0,
-        }
-    }
-
-    /// Returns the sender that can communicate with this task.
-    pub(super) fn actions_tx(&self) -> Sender<PrewarmTaskEvent> {
-        self.actions_tx.clone()
+        )
     }
 
     /// Spawns all pending transactions as blocking tasks by first chunking them.
-    fn spawn_all(&mut self) {
-        let chunk_size = (self.pending.len() / self.max_concurrency).max(1);
+    fn spawn_all(
+        &self,
+        pending: mpsc::Receiver<impl OwnedExecutableTxFor<Evm>>,
+        actions_tx: Sender<PrewarmTaskEvent>,
+    ) {
+        let executor = self.executor.clone();
+        let ctx = self.ctx.clone();
+        let max_concurrency = self.max_concurrency;
+
+        self.executor.spawn_blocking(move || {
+            let mut handles = Vec::new();
+            let (done_tx, done_rx) = mpsc::channel();
+            let mut executing = 0;
+            while let Ok(executable) = pending.recv() {
+                let task_idx = executing % max_concurrency;
+
+                if handles.len() <= task_idx {
+                    let (tx, rx) = mpsc::channel();
+                    let sender = actions_tx.clone();
+                    let ctx = ctx.clone();
+                    let done_tx = done_tx.clone();
+
+                    executor.spawn_blocking(move || {
+                        ctx.transact_batch(rx, sender, done_tx);
+                    });
+
+                    handles.push(tx);
+                }
 
-        for chunk in &self.pending.drain(..).chunks(chunk_size) {
-            let sender = self.actions_tx.clone();
-            let ctx = self.ctx.clone();
-            let pending_chunk = chunk.collect::<Vec<_>>();
+                let _ = handles[task_idx].send(executable);
 
-            self.prewarm_outcomes_left += pending_chunk.len();
-            self.executor.spawn_blocking(move || {
-                ctx.transact_batch(&pending_chunk, sender);
-            });
-        }
+                executing += 1;
+            }
+
+            // drop handle and wait for all tasks to finish and drop theirs
+            // todo: do we even need this? tasks are spawned anyway and will see all transactions
```

Comment:
this is fine, we can remove the todo here

we can perhaps get rid of this outer spawn entirely if we track the exit condition differently, and keep track of the finished execution handles at top level

but we can do this in a potential followup